### PR TITLE
Cleanup ml

### PR DESCRIPTION
**Pull request by Hugemiler**
[WIP] So @kescobo  can check the status of the cleanup

### CODE REVIEW

File: src/prediction.jl [hunk 1]
```diff
@@ -1,35 +1,68 @@
...
+# 1. Preprocessing/Wrangling Functions #
+########################################
 
-    return(longdata_df)
-    
+Resonance.dropmissing(vv) = vv[.!(ismissing.(vv))]
+dropnan(vv) = vv[.!(isnan.(vv))]
+nonna_mean(vv) = mean(dropnan(vv))
+nonmissing_mean(vv) = mean(dropmissing(vv))
+nonna_nonmissing_mean(vv) = mean(dropnan(dropmissing(vv)))
```
Comment (kescobo) [inline src/prediction.jl:54]: Not a fan of this. It seems like it doesn't save a lot of typing to just do this at the call site. If it were doing something special on a type that we're introducing, fine, but `nonna_nonmissing_mean` vs `mean(dropnan(dropmissing(` doesn't add much IMO.

File: src/prediction.jl [hunk 2]
```diff
@@ -149,44 +177,321 @@ function unstack_techreplicates(raw_data, retain_first=true)
...
+
+function train_randomforest(
+    type::Regression,
+    ref_name::String,
+    original_df,
+    input_cols,
+    output_col;
+    n_splits = 2,
+    data_preparation_fun = identity,
+    tuning_space = (
```
Comment (kescobo) [inline src/prediction.jl:195]:
```suggestion
    tuning_space = (;
```

Just making NamedTuple explicit

**Review by kescobo — COMMENTED** [review]
This branch is getting seriously hairy - can we rebase / condense any of it and get it merged? Ideally in pieces.

File: src/Resonance.jl [hunk 3]
```diff
@@ -45,20 +44,51 @@ export loadings,
...
+        Classification,
+        Regression,
+        ResonancePredictor,
+        ResonanceUnivariatePredictor,
+        ResonanceMultivariatePredictor,
+        UnivariateRandomForestClassifier,
+        UnivariateRandomForestRegressor,
+        UnivariatePredictorEnsemble,
+        # preprocessing functions
+        dropmissing,
```
[the remark this answers is not in the document]
Comment (Hugemiler) [reply src/Resonance.jl:58]: This is a different method. `DataFrames.dropmissing(df::DataFrame)` is one method, `Resonance.dropmissing(vv::AbstractArray)` is another method. This diff does not reflect the typo, though, but refer to the updated version on the new diff-to-be. Also, we can not export it, it's just a utility function to save the hassle of doing `vv[.!(ismissing.(vv))]`

File: src/Resonance.jl [hunk 4]
```diff
@@ -45,20 +44,51 @@ export loadings,
...
-       confmatrix2barplot,
-       regression_bestprediction,
-       non_na_mean
+export  # structs/types
+        Prediction,
+        Classification,
+        Regression,
+        ResonancePredictor,
+        ResonanceUnivariatePredictor,
+        ResonanceMultivariatePredictor,
```
[the remark this answers is not in the document]
Comment (Hugemiler) [reply src/Resonance.jl:53]: Yes. Done.

File: src/prediction.jl (hunk 2 above)
Comment (Hugemiler) [reply src/prediction.jl:320]: I didn't know that the `NamedTuple` constructor worked like that (names are actually keyword arguments). I will alter the occurrences of the tuning_space accordingly. Thank you.

File: src/prediction.jl (hunk 1 above)
Comment (Hugemiler) [reply src/prediction.jl:57]: Resolved.

**Review by Hugemiler — COMMENTED** [review]
Switching to another branch with only source changes

**Comment by kescobo** [comment]
Let's link those other PRs

- Stuff in `src/`: #75
- Notebooks: TBD
- Runnable script: TBD

**Pull request closed without merging**